sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -91,10 +91,15 @@ case class AdaptiveSparkPlanExec(
     DisableUnnecessaryBucketedScan
   ) ++ context.session.sessionState.queryStagePrepRules
 
+  @transient private val initialPlan = context.session.withActive {
+    applyPhysicalRules(
+      inputPlan, queryStagePreparationRules, Some((planChangeLogger, "AQE Preparations")))
+  }
+
   // A list of physical optimizer rules to be applied to a new stage before its execution. These
   // optimizations should be stage-independent.
   @transient private val queryStageOptimizerRules: Seq[Rule[SparkPlan]] = Seq(
-    PlanAdaptiveDynamicPruningFilters(context.stageCache),
+    PlanAdaptiveDynamicPruningFilters(initialPlan),
```

Comment:
We need use the `initialPlan `not the `inputPlan`, because the `inputPlan `is not applied the `queryStagePreparationRules(EnsureRequirements)`.

Reply:
I think it's better to pass `this` as the root plan. `AdaptiveSparkPlanExec` keeps changing when more and more query stages are completed. So it's better that `PlanAdaptiveDynamicPruningFilters` always look at the latest plan.